📝 add instructions for setting up vscode IDE (#4190)
OctoPrint/OctoPrint: docs/development/environment.rst

   Make sure you are running a PyCharm version of 2016.1 or later, or manually fix
   `a debugger bug contained in earlier versions <https://youtrack.jetbrains.com/issue/PY-18365>`_ or plugin management
   will not work in your developer install when running OctoPrint from PyCharm in debug mode.

Visual Studio Code (vscode)
---------------------------

>   - Install Visual Studio from https://code.visualstudio.com/Download

Reviewer: ```suggestion
  - Install Visual Studio Code from https://code.visualstudio.com/Download
```

  - Open folder select OctoPrint checkout folder (e.g. ``~/devel/OctoPrint`` or ``C:\Devel\OctoPrint``)

>     settings.json

Reviewer: Doesn't make sense here
```suggestion
```

            "python.formatting.blackArgs": [
                "--config",
                "black.toml"
            ],
            "editor.formatOnSave": true,
            "python.sortImports.args": [
                "--profile=black",
            ],
            "[python]": {
                "editor.codeActionsOnSave": {
                    "source.organizeImports": true
                }
            },
            "python.linting.pylintEnabled": false,
            "python.linting.flake8Enabled": true,
            "python.linting.enabled": true
        }

    tasks.json
      .. code-block:: json

        {
          "version": "2.0.0",
          "tasks": [
            {
                "label": "clean build artifacts",
                "type": "shell",
                "command": "python ./setup.py clean"
            },
            {
                "label": "build docs",
                "type": "shell",
                "command": "sphinx-build -b html ./docs ./docs/_build"
            }
          ]
        }

    launch.json
      .. code-block:: json

        {
          "version": "0.2.0",
          "configurations": [
              {

>                   "name": "Octoprint",

Reviewer: ```suggestion
                  "name": "OctoPrint",
```
Very picky this one... 🙂

                  "type": "python",
                  "request": "launch",
                  "module": "octoprint",
                  "args": [
                      "serve",
                      "--debug"
                  ],
                  "cwd": "${workspaceFolder}/src",
                  "preLaunchTask": "clean build artifacts"
              }
          ]
        }
  
  In the terminal install the python extension by running this command:
    .. code-block:: bash

      code --install-extension ms-python.python

  In vscode terminal, or with venv active install code formatter black and linter flake8 by running:
    .. code-block:: bash

      python -m pip install -U black flake8 flake8-bugbear

  Summary of vscode config:

  * Pressing f5 will now start OctoPrint in debug

>   * Your terminal inside vscode uses the virtrual python environment

Reviewer: ```suggestion
  * Your terminal inside vscode uses the virtual python environment
```

   Make sure you are running a PyCharm version of 2016.1 or later, or manually fix
   `a debugger bug contained in earlier versions <https://youtrack.jetbrains.com/issue/PY-18365>`_ or plugin management
   will not work in your developer install when running OctoPrint from PyCharm in debug mode.

Visual Studio Code (vscode)
---------------------------

>   - Install Visual Studio Code from https://code.visualstudio.com/Download

Reviewer 2: Somehow the missing newline here irks me 😅

```suggestion

  - Install Visual Studio Code from `code.visualstudio.com <https://code.visualstudio.com/Download>`_
```

  - Open folder select OctoPrint checkout folder (e.g. ``~/devel/OctoPrint`` or ``C:\Devel\OctoPrint``)

>   - Create a directory '.vscode' if not already present in the root of the project

Reviewer 2: ```suggestion
  - Create a directory ``.vscode`` if not already present in the root of the project
```

>   - Create the following files inside the .vscode directory

Reviewer 2: ```suggestion
  - Create the following files inside the ``.vscode`` directory
```

    settings.json

>       .. code-block:: json
> 
>         {
>             "python.defaultInterpreterPath": "venv3/bin/python",
>             "python.formatting.provider": "black",
>             "python.formatting.blackArgs": [
>                 "--config",
>                 "black.toml"
>             ],
>             "editor.formatOnSave": true,
>             "python.sortImports.args": [
>                 "--profile=black",
>             ],
>             "[python]": {
>                 "editor.codeActionsOnSave": {
>                     "source.organizeImports": true
>                 }
>             },
>             "python.linting.pylintEnabled": false,
>             "python.linting.flake8Enabled": true,
>             "python.linting.enabled": true
>         }

Reviewer 2: Something's not right with that indent...

```suggestion
      .. code-block:: json

         {
             "python.defaultInterpreterPath": "venv3/bin/python",
             "python.formatting.provider": "black",
             "python.formatting.blackArgs": [
                 "--config",
                 "black.toml"
             ],
             "editor.formatOnSave": true,
             "python.sortImports.args": [
                 "--profile=black",
             ],
             "[python]": {
                 "editor.codeActionsOnSave": {
                     "source.organizeImports": true
                 }
             },
             "python.linting.pylintEnabled": false,
             "python.linting.flake8Enabled": true,
             "python.linting.enabled": true
         }
```

    tasks.json

>       .. code-block:: json
> 
>         {
>           "version": "2.0.0",
>           "tasks": [
>             {
>                 "label": "clean build artifacts",
>                 "type": "shell",
>                 "command": "python ./setup.py clean"
>             },
>             {
>                 "label": "build docs",
>                 "type": "shell",
>                 "command": "sphinx-build -b html ./docs ./docs/_build"
>             }
>           ]
>         }

Reviewer 2: Indent

```suggestion
      .. code-block:: json

         {
           "version": "2.0.0",
           "tasks": [
             {
                 "label": "clean build artifacts",
                 "type": "shell",
                 "command": "python ./setup.py clean"
             },
             {
                 "label": "build docs",
                 "type": "shell",
                 "command": "sphinx-build -b html ./docs ./docs/_build"
             }
           ]
         }
```

    launch.json

>       .. code-block:: json
> 
>         {
>           "version": "0.2.0",
>           "configurations": [
>               {
>                   "name": "OctoPrint",
>                   "type": "python",
>                   "request": "launch",
>                   "module": "octoprint",
>                   "args": [
>                       "serve",
>                       "--debug"
>                   ],
>                   "cwd": "${workspaceFolder}/src",
>                   "preLaunchTask": "clean build artifacts"
>               }
>           ]
>         }

Reviewer 2: Indent

```suggestion
      .. code-block:: json

         {
           "version": "0.2.0",
           "configurations": [
               {
                   "name": "OctoPrint",
                   "type": "python",
                   "request": "launch",
                   "module": "octoprint",
                   "args": [
                       "serve",
                       "--debug"
                   ],
                   "cwd": "${workspaceFolder}/src",
                   "preLaunchTask": "clean build artifacts"
               }
           ]
         }
```

  
  In the terminal install the python extension by running this command:

>     .. code-block:: bash

Reviewer 2: Newline

```suggestion

    .. code-block:: bash
```

      code --install-extension ms-python.python

  In vscode terminal, or with venv active install code formatter black and linter flake8 by running:

>     .. code-block:: bash

Reviewer 2: Newline

```suggestion

    .. code-block:: bash
```

      python -m pip install -U black flake8 flake8-bugbear

  Summary of vscode config:

>   * Pressing f5 will now start OctoPrint in debug

Reviewer 2: ```suggestion
  * Pressing ``F5`` will now start OctoPrint in debug mode
```

  * Your terminal inside vscode uses the virtual python environment

  * Saving a file will run an auto formatter and import sort
  

>   * Ctrl + shift + b can be used to run the 'build docs' task to rebuild the documentation

Reviewer 2: ```suggestion
  * ``Ctrl+Shift+B`` can be used to run the ``build docs`` task to rebuild the documentation
```
